File: 0017-letter-combinations-of-a-phone-number/depth-first-search.py
class Solution:
    def dfs(self, index, digits, hmap, tmp, res):
        if (index == len(digits)):
            res.append(tmp)
            return
        else:
            for i in hmap[digits[index]]:
                self.dfs(index + 1, digits, hmap, tmp + i, res)

    def letterCombinations(self, digits: str) -> List[str]:
        result = []
        HMAP = { "2": "abc", "3": "def", "4": "ghi", "5": "jkl", "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"}

        if (len(digits) != 0):
            self.dfs(0, digits, HMAP, "", result)
        
        return result
        

File: 0017-letter-combinations-of-a-phone-number/depth-first-search.cpp
using namespace std;

class Solution 
{
    public:
        vector<string> letterCombinations(string digits)
        {
            vector<string> result;
            if (!digits.empty())
            {
                dfs(0, digits, "", result);
            }

            return result;
        }

    private:
        const vector<string> letters{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

        void dfs(int i, const string& digits, string&& tmp, vector<string>& res)
        {
            if (i == digits.length())
            {
                res.push_back(tmp);
                return;
            }
            else
            {
                for (const char letter : letters[digits[i] - '0'])
                {
                    tmp.push_back(letter);
                    dfs(i + 1, digits, move(tmp), res);
                    tmp.pop_back();
                }
            }
        }
};